src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs

```diff
@@ -22,11 +23,11 @@ public abstract class HttpResponse
 
         public abstract IResponseCookies Cookies { get; }
 
-        public abstract bool HeadersSent { get; }
+        public abstract bool HasStarted { get; }
 
-        public abstract void OnSendingHeaders(Action<object> callback, object state);
+        public abstract void OnResponseStarting(Func<object, Task> callback, object state);
```

Comment:
Yeah, does that mean OSH callbacks will be async only? :smile: 
Is it worth re-opening the ticket in aspnet/Security then?